diff --git a/tools/external_ic.F90 b/tools/external_ic.F90
index 4c06938..4e25964 100644
--- a/tools/external_ic.F90
+++ b/tools/external_ic.F90
@@ -593,7 +593,7 @@ contains
       call register_axis(SFC_restart, dim_names_alloc(2), "y")
       call register_axis(SFC_restart, dim_names_alloc(1), "x")
       call register_restart_field(SFC_restart, 'tsea', Atm%ts, dim_names_alloc)
-      call read_restart(SFC_restart)
+      call read_restart(SFC_restart,ignore_checksum=.true.)
       call close_file(SFC_restart)
       deallocate (dim_names_alloc)
     else
@@ -632,7 +632,7 @@ contains
         ! land-frac
         call register_restart_field(ORO_restart, 'land_frac', Atm%oro, dim_names_2d)
       endif
-      call read_restart(ORO_restart)
+      call read_restart(ORO_restart,ignore_checksum=.true.)
       call close_file(ORO_restart)
     else
       call mpp_error(FATAL,'==> Error in External_ic::get_nggps_ic: tiled file '//trim(fn_oro_ics)//' for NGGPS IC does not exist')
@@ -952,7 +952,7 @@ contains
           enddo
 
           ! read in the gfs_data and free the restart type to be re-used by the nest
-          call read_restart(GFS_restart)
+          call read_restart(GFS_restart,ignore_checksum=.true.)
           call close_file(GFS_restart)
         else
           call mpp_error(FATAL,'==> Error in External_ic::get_nggps_ic: tiled file '//trim(fn_gfs_ics)//' for NGGPS IC does not exist')
@@ -1109,7 +1109,7 @@ contains
           call register_axis(SFC_restart, dim_names_2d(2), "y")
           call register_axis(SFC_restart, dim_names_2d(1), "x")
           call register_restart_field(SFC_restart, 'tsea', Atm%ts, dim_names_2d)
-          call read_restart(SFC_restart)
+          call read_restart(SFC_restart,ignore_checksum=.true.)
           call close_file(SFC_restart)
         else
           call mpp_error(FATAL,'==> Error in External_ic::get_hrrr_ic: tiled file '//trim(fn_sfc_ics)//' for HRRR IC does not exist')
@@ -1144,7 +1144,7 @@ contains
             ! land-frac
             call register_restart_field(ORO_restart, 'land_frac', Atm%oro, (/"lat", "lon"/))
           endif
-          call read_restart(ORO_restart)
+          call read_restart(ORO_restart,ignore_checksum=.true.)
           call close_file(ORO_restart)
         else
           call mpp_error(FATAL,'==> Error in External_ic::get_hrrr_ic: tiled file '//trim(fn_oro_ics)//' for HRRR IC does not exist')
@@ -1193,7 +1193,7 @@ contains
             call get_tracer_names(MODEL_ATMOS, nt, tracer_name)
             call register_restart_field(HRRR_restart, trim(tracer_name), q(:,:,:,nt), dim_names_3d3, is_optional=.true.)
           enddo
-          call read_restart(HRRR_restart)
+          call read_restart(HRRR_restart,ignore_checksum=.true.)
           call close_file(HRRR_restart)
         else
           call mpp_error(FATAL,'==> Error in External_ic::get_hrrr_ic: tiled file '//trim(fn_hrr_ics)//' for HRRR IC does not exist')
@@ -2068,7 +2068,7 @@ contains
           elseif (.not. filtered_terrain) then
             call register_restart_field(ORO_restart, 'orog_raw', Atm%phis, dim_names_2d)
         endif
-        call read_restart(ORO_restart)
+        call read_restart(ORO_restart,ignore_checksum=.true.)
         call close_file(ORO_restart)
       endif
       Atm%phis = Atm%phis*grav
@@ -2100,7 +2100,7 @@ contains
 #endif
         call register_restart_field(GFS_restart, 'ps', ps_gfs, dim_names_2d)
         call register_restart_field(GFS_restart, 'ZH', zh_gfs, dim_names_3d4)
-        call read_restart(GFS_restart)
+        call read_restart(GFS_restart,ignore_checksum=.true.)
         call close_file(GFS_restart)
       endif
 
